extensions/azure/processors/PutAzureDataLakeStorage.cpp

```diff
@@ -0,0 +1,167 @@
+
+  if (!context->getProperty(FilesystemName, params.file_system_name, flow_file) || params.file_system_name.empty()) {
+    logger_->log_error("Filesystem Name '%s' is invalid or empty!", params.file_system_name);
+    return std::nullopt;
+  }
+
+  context->getProperty(DirectoryName, params.directory_name, flow_file);
+
+  context->getProperty(FileName, params.filename, flow_file);
+  if (params.filename.empty() && (!flow_file->getAttribute("filename", params.filename) || params.filename.empty())) {
+    logger_->log_error("No File Name is set and default object key 'filename' attribute could not be found!");
+    return std::nullopt;
+  }
+
+  return params;
+}
+
+void PutAzureDataLakeStorage::onTrigger(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSession>& session) {
+  logger_->log_debug("PutAzureDataLakeStorage onTrigger");
+  std::shared_ptr<core::FlowFile> flow_file = session->get();
+  if (!flow_file) {
+    context->yield();
+    return;
+  }
+
+  const auto params = buildUploadParameters(context, flow_file);
+  if (!params) {
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  PutAzureDataLakeStorage::ReadCallback callback(flow_file->getSize(), azure_data_lake_storage_, *params, logger_);
+  session->read(flow_file, &callback);
+  auto result = callback.getResult();
+  if (result.result_code == storage::UploadResultCode::FILE_ALREADY_EXISTS) {
+    gsl_Expects(conflict_resolution_strategy_ != FileExistsResolutionStrategy::REPLACE_FILE);
+    if (conflict_resolution_strategy_ == FileExistsResolutionStrategy::FAIL_FLOW) {
+      logger_->log_error("Failed to upload file '%s/%s' to filesystem '%s' on Azure Data Lake storage because file already exists",
+        params->directory_name, params->filename, params->file_system_name);
+      session->transfer(flow_file, Failure);
+      return;
+    } else if (conflict_resolution_strategy_ == FileExistsResolutionStrategy::IGNORE_REQUEST) {
+      logger_->log_debug("Upload of file '%s/%s' was ignored because it already exits in filesystem '%s' on Azure Data Lake Storage",
+        params->directory_name, params->filename, params->file_system_name);
+      session->transfer(flow_file, Success);
+      return;
+    }
+  } else if (result.result_code == storage::UploadResultCode::FAILURE) {
+    logger_->log_error("Failed to upload file '%s/%s' to filesystem '%s' on Azure Data Lake storage", params->directory_name, params->filename, params->file_system_name);
+    session->transfer(flow_file, Failure);
+  } else {
+    session->putAttribute(flow_file, "azure.filesystem", params->file_system_name);
+    session->putAttribute(flow_file, "azure.directory", params->directory_name);
+    session->putAttribute(flow_file, "azure.filename", params->filename);
+    session->putAttribute(flow_file, "azure.primaryUri", result.primary_uri);
+    session->putAttribute(flow_file, "azure.length", std::to_string(result.length));
```

Comment:
Is it useful to have this `length` field in the `result` struct?  Can its value be different from `flow_file->getSize()`?

Reply:
I suppose we do not need it, especially as it is part of the input. Updated in c8257e48b454819ecef4cc2f527f2457a5f81b89